<a href="https://colab.research.google.com/github/bakut/packt-data-science-workshop/blob/master/Tuning_n_estimators_to_Reduce_Overfitting.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Tuning n_estimators to Reduce Overfitting**

Train a Random Forest classifier to predict the type of an animal based on its attributes and will try two different values for the n_estimators hyperparameter:

In [0]:
#importing packages
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

In [0]:
#create a variable called file_url that contains url dataset
file_url = 'https://raw.githubusercontent.com/PacktWorkshops/The-Data-Science-Workshop/master/Chapter04/Dataset/openml_phpZNNasq.csv'

In [0]:
#load the dataset into a dataframe
df = pd.read_csv(file_url)

In [0]:
#remove the animal colum using .drop() and extract the type target variable into a new variable called y using .pop()
df.drop(columns='animal', inplace=True)
y = df.pop('type')

In [0]:
#split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(df, y, test_size=0.4, random_state=188)

In [8]:
#instantiate RandomForestClassifier
rf_model = RandomForestClassifier(random_state=42, n_estimators=1)
rf_model.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=1,
                       n_jobs=None, oob_score=False, random_state=42, verbose=0,
                       warm_start=False)

In [0]:
#Make predictions on the training and testing sets
train_preds = rf_model.predict(X_train)
test_preds = rf_model.predict(X_test)

In [0]:
#calculate the accuracy score for the training and testing sets
train_acc = accuracy_score(y_train, train_preds)
test_acc = accuracy_score(y_test, test_preds)

In [11]:
#print the accuracy scores
print(train_acc)
print(test_acc)

0.9166666666666666
0.8048780487804879


The accuracy score decreased for both the training and testing sets. But now the difference is smaller compared to the results from Exercise 4.01.


In [12]:
#instantiate another RandomForestClassifier with random_state=42 and n_estimators=30, and then fit the model with the training set
rf_model2 = RandomForestClassifier(random_state=42, n_estimators=30)
rf_model2.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=30,
                       n_jobs=None, oob_score=False, random_state=42, verbose=0,
                       warm_start=False)

In [0]:
#make predictions on the training and testing sets with .predict() and ssave results into two new variables
train_preds2 = rf_model2.predict(X_train)
test_preds2 = rf_model2.predict(X_test)

In [0]:
#calculate the accuracy score for the training and testing sets and save the result in two new variables
train_acc2 = accuracy_score(y_train, train_preds2)
test_acc2 = accuracy_score(y_test, test_preds2)

In [18]:
#print the accuracy scores
print(train_acc2)
print(test_acc2)

1.0
0.9024390243902439


This output shows us the model is overfitting less compared to the results from the previous step and still has a very high-performance level for the training set.

In the previous exercise, we achieved an accuracy score of 1 for the training set and 0.88 for the testing one. In this exercise, we trained two additional Random Forest models with n_estimators = 3 and 30. The model with the lowest number of trees has the lowest accuracy: 0.92 (training) and 0.8 (testing). On the other hand, increasing the number of trees to 30, we achieved a higher accuracy: 1 and 0.9. Our model is overfitting slightly less now. It is not perfect, but it is a good start.